libs/go/aealite/protocols/dialogue.go

```diff
@@ -0,0 +1,437 @@
+package protocols
+
+import (
+	"crypto/rand"
+	"encoding/hex"
+	"errors"
+	"fmt"
+)
+
+type Address string
+type Performative []string
+type SomeMessageType string
+
+const (
+	NONCE_BYTES_NB = 32
+)
+
+type Role string
+
+const (
+	Role1             Role = "role1"
+	Role2             Role = "role2"
+	StartingMessageId      = 1
+	StartingTarget         = 0
+)
+
+type RuleType struct {
+	initialPerformatives  Performative
+	terminalPerformatives Performative
+	validReplies map[Performative][]Performative
+}
+
+type MessageId int
+type Target int
+type DialogueInterface interface {
+	getSelfAddress() Address
+	getRole() Role
+	getRules() RuleType
+	getMessageClass() AbstractMessage
+	isSelfInitiated() bool
+	getLastIncomingMessage() AbstractMessage
+	getLastOutgoingMessage() AbstractMessage
+	getLastMessage() AbstractMessage
+	counterPartyForMessage(AbstractMessage) bool
+	isMessageBySelf(AbstractMessage) bool
+	isMessageByOther(AbstractMessage) bool
+	getMessage(MessageId) AbstractMessage
+	hasMessageId(MessageId) bool
+
+	update(AbstractMessage)
+	validateNextMessage(AbstractMessage) (bool, string)
+	basicValidations(AbstractMessage) (bool, string)
+	basicValidationInitialMessage(AbstractMessage) (bool, string)
+	basicValidationNonInitialMessage(AbstractMessage) (bool, string)
```

Comment:
why `(bool, string)` and not `error`? Then, if `error_obj == nil` means validation succeded, else `error_obj.Error()` returns the error message.

Reply:
Now I see; this comes from Python code. Let us leave this as it is, although we could think of a custom struct `ValidationResult` to better encapsulate the returned value of these methods.

Reply:
I think the python code is no good guide, I prefer the error to be returned as Marco suggests, that's more idiomatic. But can be done later of course.